src/EditorFeatures/Core.Wpf/InlineHints/InlineHintsKeyProcessorProvider.cs

```diff
@@ -0,0 +1,141 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.ComponentModel.Composition;
+using System.Linq;
+using System.Runtime.CompilerServices;
+using System.Windows.Input;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.InlineHints;
+using Microsoft.CodeAnalysis.Options;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.VisualStudio.Text.Editor;
+using Microsoft.VisualStudio.Utilities;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Editor.InlineHints
+{
+    /// <summary>
+    /// Key processor that allows us to toggle inline hints when a user hits ctrl-alt.
+    /// </summary>
+    [Export(typeof(IKeyProcessorProvider))]
+    [TextViewRole(PredefinedTextViewRoles.Document)]
+    [ContentType(ContentTypeNames.RoslynContentType)]
+    [Name(nameof(InlineHintsKeyProcessorProvider))]
+    [Order(Before = "default")]
+    internal class InlineHintsKeyProcessorProvider : IKeyProcessorProvider
+    {
+        private static readonly ConditionalWeakTable<IWpfTextView, InlineHintsKeyProcessor> s_viewToProcessor = new();
+
+        private readonly IGlobalOptionService _globalOptionService;
+
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public InlineHintsKeyProcessorProvider(IGlobalOptionService globalOptionService)
+        {
+            _globalOptionService = globalOptionService;
+        }
+
+        public KeyProcessor GetAssociatedProcessor(IWpfTextView wpfTextView)
+        {
+            return s_viewToProcessor.GetValue(wpfTextView, v => new InlineHintsKeyProcessor(_globalOptionService, v));
+        }
+
+        private class InlineHintsKeyProcessor : KeyProcessor
+        {
+            private readonly IGlobalOptionService _globalOptionService;
+            private readonly IWpfTextView _view;
+
+            public InlineHintsKeyProcessor(IGlobalOptionService globalOptionService, IWpfTextView view)
+            {
+                _globalOptionService = globalOptionService;
+                _view = view;
+                _view.Closed += OnViewClosed;
+                _view.LostAggregateFocus += OnLostFocus;
+            }
+
+            private static bool IsCtrlOrAlt(KeyEventArgs args)
+                => args.Key is Key.LeftCtrl or Key.RightCtrl or Key.LeftAlt or Key.RightAlt;
+
+            private Document? GetDocument()
+            {
+                var document =
+                    _view.BufferGraph.GetTextBuffers(b => true)
+                                     .Select(b => b.CurrentSnapshot.GetOpenDocumentInCurrentContextWithChanges())
+                                     .WhereNotNull()
+                                     .FirstOrDefault();
+
+                return document;
+            }
```

Comment:
Do we really need to check this? I can't really come up with a case where our content type is set but we don't have a document in a way we'd really care about here. More a "less code is better" here.

Reply:
(general concern also being as this PR stands this is running on every key event and every modifier...)